// clarity-angular/stack-view/stack-view.js
import { Component, Directive, EventEmitter, Output } from "@angular/core";
export var StackView = (function () {
    function StackView() {
        /**
         * Undocumented experimental feature: inline editing.
         */
        this.editable = false;
        this.save = new EventEmitter(false);
        this._editMode = false;
        this.editingChange = new EventEmitter(false);
    }
    Object.defineProperty(StackView.prototype, "editing", {
        get: function () {
            return this.editable && this._editMode;
        },
        set: function (value) {
            if (this.editable) {
                this._editMode = value;
                this.editingChange.emit(value);
                if (!value) {
                    this.save.emit(null);
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * End of undocumented experimental feature.
     */
    StackView.decorators = [
        { type: Component, args: [{
                    selector: "clr-stack-view",
                    template: "\n        <ng-content select=\"clr-stack-header\"></ng-content>\n        <dl class=\"stack-view\"><ng-content></ng-content></dl>\n    ",
                    // Custom elements are inline by default.
                    styles: ["\n        :host { display: block; }\n    "]
                },] },
    ];
    /** @nocollapse */
    StackView.ctorParameters = function () { return []; };
    StackView.propDecorators = {
        'save': [{ type: Output, args: ["clrStackSave",] },],
    };
    return StackView;
}());
export var StackViewCustomTags = (function () {
    function StackViewCustomTags() {
    }
    // No behavior
    // The only purpose is to "declare" the tag in Angular2
    StackViewCustomTags.decorators = [
        { type: Directive, args: [{ selector: "clr-stack-label, clr-stack-content" },] },
    ];
    /** @nocollapse */
    StackViewCustomTags.ctorParameters = function () { return []; };
    return StackViewCustomTags;
}());
