// clarity-angular/datagrid/datagrid-action-overflow.js
import { Component, EventEmitter, HostListener, Input, Output, ElementRef, ViewChildren } from "@angular/core";
import { RowActionService } from "./providers/row-action-service";
import { Popover, Direction } from "../popover/popover";
export var DatagridActionOverflow = (function () {
    function DatagridActionOverflow(elementRef, rowActionService) {
        this.elementRef = elementRef;
        this.rowActionService = rowActionService;
        /**
         * Tracks whether the action overflow menu is open or not
         */
        this._open = false;
        this.openChanged = new EventEmitter(false);
    }
    DatagridActionOverflow.prototype.ngAfterViewInit = function () {
        var _this = this;
        this._menuSubscription = this.menu.changes.subscribe(function () {
            if (_this.menu.length > 0) {
                _this.rowActionService.open(function () {
                    _this.position = new Popover(_this.menu.first.nativeElement);
                    _this.position.anchor(_this.elementRef.nativeElement, Direction.RIGHT, { userAnchorParent: true });
                });
            }
            else {
                _this.position.destroy();
                delete _this.position;
                _this.rowActionService.close();
            }
        });
    };
    DatagridActionOverflow.prototype.ngOnDestroy = function () {
        this._menuSubscription.unsubscribe();
    };
    Object.defineProperty(DatagridActionOverflow.prototype, "open", {
        get: function () {
            return this._open;
        },
        set: function (open) {
            var boolOpen = !!open;
            if (boolOpen !== this._open) {
                this._open = boolOpen;
                this.openChanged.emit(boolOpen);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Shows/hides the action overflow menu
     */
    DatagridActionOverflow.prototype.toggle = function () {
        this.open = !this.open;
    };
    //called on mouse clicks anywhere in the DOM.
    //Checks to see if the mouseclick happened on the host or outside
    DatagridActionOverflow.prototype.onMouseClick = function (target) {
        if (this._open) {
            var current = target; //Get the element in the DOM on which the mouse was clicked
            var actionMenuHost = this.elementRef.nativeElement; //Get the current actionMenu native HTML element
            if (target.className === "datagrid-action-overflow") {
                return; // if clicking on the action overflow container but not the content, return without closing
            }
            //Start checking if current and actionMenuHost are equal. If not traverse to the parentNode and check again.
            while (current) {
                if (current.className === "datagrid-action-overflow") {
                    break; // if user clicked on the overflow menu, hide it
                }
                if (current === actionMenuHost) {
                    return;
                }
                current = current.parentNode;
            }
            this._open = false; // Hide the overflow menu
        }
    };
    DatagridActionOverflow.decorators = [
        { type: Component, args: [{
                    selector: "clr-dg-action-overflow",
                    template: "\n        <clr-icon shape=\"ellipsis-vertical\" class=\"datagrid-action-toggle\" (click)=\"toggle()\"></clr-icon>\n        <div #menu class=\"datagrid-action-overflow\" *ngIf=\"open\">\n            <ng-content></ng-content>\n        </div>\n    "
                },] },
    ];
    /** @nocollapse */
    DatagridActionOverflow.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: RowActionService, },
    ]; };
    DatagridActionOverflow.propDecorators = {
        'menu': [{ type: ViewChildren, args: ["menu",] },],
        'open': [{ type: Input, args: ["clrDgActionOverflowOpen",] },],
        'openChanged': [{ type: Output, args: ["clrDgActionOverflowOpenChange",] },],
        'onMouseClick': [{ type: HostListener, args: ["document:click", ["$event.target"],] },],
    };
    return DatagridActionOverflow;
}());
