// clarity-angular/datagrid/render/body-renderer.js
import { Directive, ElementRef } from "@angular/core";
import { DatagridRenderOrganizer } from "./render-organizer";
import { DomAdapter } from "./dom-adapter";
export var DatagridBodyRenderer = (function () {
    function DatagridBodyRenderer(el, organizer, domAdapter) {
        var _this = this;
        this.el = el;
        this.organizer = organizer;
        this.domAdapter = domAdapter;
        this.subscription = organizer.scrollbar.subscribe(function () { return _this.computeScrollbarWidth(); });
    }
    DatagridBodyRenderer.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    DatagridBodyRenderer.prototype.computeScrollbarWidth = function () {
        this.organizer.scrollbarWidth.next(this.domAdapter.scrollBarWidth(this.el.nativeElement));
    };
    DatagridBodyRenderer.decorators = [
        { type: Directive, args: [{
                    selector: "[clrDgBody]"
                },] },
    ];
    /** @nocollapse */
    DatagridBodyRenderer.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: DatagridRenderOrganizer, },
        { type: DomAdapter, },
    ]; };
    return DatagridBodyRenderer;
}());
