// clarity-angular/datagrid/render/cell-renderer.js
import { Directive, ElementRef, Renderer } from "@angular/core";
import { STRICT_WIDTH_CLASS } from "./constants";
import { DatagridRenderOrganizer } from "./render-organizer";
export var DatagridCellRenderer = (function () {
    function DatagridCellRenderer(el, renderer, organizer) {
        var _this = this;
        this.el = el;
        this.renderer = renderer;
        this.subscription = organizer.clearWidths.subscribe(function () { return _this.clearWidth(); });
    }
    DatagridCellRenderer.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    DatagridCellRenderer.prototype.clearWidth = function () {
        this.renderer.setElementClass(this.el.nativeElement, STRICT_WIDTH_CLASS, false);
        this.renderer.setElementStyle(this.el.nativeElement, "width", null);
    };
    DatagridCellRenderer.prototype.setWidth = function (strict, value) {
        this.renderer.setElementClass(this.el.nativeElement, STRICT_WIDTH_CLASS, strict);
        this.renderer.setElementStyle(this.el.nativeElement, "width", value + "px");
    };
    DatagridCellRenderer.decorators = [
        { type: Directive, args: [{
                    selector: "clr-dg-cell"
                },] },
    ];
    /** @nocollapse */
    DatagridCellRenderer.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer, },
        { type: DatagridRenderOrganizer, },
    ]; };
    return DatagridCellRenderer;
}());
