// clarity-angular/datagrid/render/head-renderer.js
import { Directive, ElementRef, Renderer } from "@angular/core";
import { DatagridRenderOrganizer } from "./render-organizer";
export var DatagridHeadRenderer = (function () {
    function DatagridHeadRenderer(el, renderer, organizer) {
        var _this = this;
        this.el = el;
        this.renderer = renderer;
        this.subscription = organizer.scrollbarWidth.subscribe(function (width) { return _this.accountForScrollbar(width); });
    }
    DatagridHeadRenderer.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    DatagridHeadRenderer.prototype.accountForScrollbar = function (width) {
        this.renderer.setElementStyle(this.el.nativeElement, "padding-right", width + "px");
    };
    DatagridHeadRenderer.decorators = [
        { type: Directive, args: [{
                    selector: "[clrDgHead]"
                },] },
    ];
    /** @nocollapse */
    DatagridHeadRenderer.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer, },
        { type: DatagridRenderOrganizer, },
    ]; };
    return DatagridHeadRenderer;
}());
