// clarity-angular/datagrid/render/table-renderer.js
import { Directive, ElementRef, Renderer } from "@angular/core";
import { COMPUTE_WIDTH_CLASS } from "./constants";
import { DatagridRenderOrganizer } from "./render-organizer";
export var DatagridTableRenderer = (function () {
    function DatagridTableRenderer(el, renderer, organizer) {
        var _this = this;
        this.el = el;
        this.renderer = renderer;
        this.subscription = organizer.tableMode.subscribe(function (on) { return _this.tableMode(on); });
    }
    DatagridTableRenderer.prototype.ngOnDestroy = function () {
        this.subscription.unsubscribe();
    };
    DatagridTableRenderer.prototype.tableMode = function (on) {
        this.renderer.setElementClass(this.el.nativeElement, COMPUTE_WIDTH_CLASS, on);
    };
    DatagridTableRenderer.decorators = [
        { type: Directive, args: [{
                    selector: "[clrDgTableWrapper]"
                },] },
    ];
    /** @nocollapse */
    DatagridTableRenderer.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer, },
        { type: DatagridRenderOrganizer, },
    ]; };
    return DatagridTableRenderer;
}());
