// clarity-angular/wizard/wizard-page.js
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
import { Component, EventEmitter, Input, Output, ViewChild } from "@angular/core";
import { TabContent } from "../tabs/tab-content";
export var WizardPage = (function (_super) {
    __extends(WizardPage, _super);
    function WizardPage() {
        _super.call(this);
        this.hasProjectedTitleContent = false;
        // User can bind his event handler for onCommit of the main content
        this.onCommit = new EventEmitter(false);
        // User can bind his/her event handler for onLoad of the main content
        this.onLoad = new EventEmitter(false);
        // input variable, optional, to set if this tab is skipped
        this.isSkipped = false;
        // input variable, optional, to set if this tab is skipped
        this.preventDefault = false;
        // Emitter for Next button state changes
        this.nextDisabledChanged = new EventEmitter(false);
    }
    WizardPage.prototype.ngOnInit = function () {
        var projectedTitleHTML = this.titleContainer.nativeElement.innerHTML.trim();
        this.hasProjectedTitleContent = projectedTitleHTML.length > 0;
    };
    WizardPage.decorators = [
        { type: Component, args: [{
                    selector: "clr-wizard-page",
                    template: "\n      <!--\n        ~ Copyright (c) 2016 VMware, Inc. All Rights Reserved.\n        ~ This software is released under MIT license.\n        ~ The full license information can be found in LICENSE in the root directory of this project.\n        -->\n\n      <section>\n          <label #titleContainer class=\"text-light\">\n              {{title}}\n              <ng-content select=\".wizard-page-title\"></ng-content>\n          </label>\n\n          <div class=\"content-wrapper\">\n              <ng-content></ng-content>\n          </div>\n      </section>\n    ",
                    host: {
                        "[id]": "id",
                        "[class.clr-nav-content]": "true",
                        "role": "tabpanel",
                        "[attr.aria-hidden]": "!active",
                        "[attr.aria-labelledby]": "ariaLabelledBy",
                        "[attr.data-hidden]": "!active",
                        "[class.active]": "active"
                    }
                },] },
    ];
    /** @nocollapse */
    WizardPage.ctorParameters = function () { return []; };
    WizardPage.propDecorators = {
        'titleContainer': [{ type: ViewChild, args: ["titleContainer",] },],
        'onCommit': [{ type: Output, args: ["clrWizardPageOnCommit",] },],
        'onLoad': [{ type: Output, args: ["clrWizardPageOnLoad",] },],
        'isSkipped': [{ type: Input, args: ["clrWizardPageIsSkipped",] },],
        'preventDefault': [{ type: Input, args: ["clrWizardPagePreventDefault",] },],
        'nextDisabled': [{ type: Input, args: ["clrWizardPageNextDisabled",] },],
        'errorFlag': [{ type: Input, args: ["clrWizardPageErrorFlag",] },],
        'nextDisabledChanged': [{ type: Output, args: ["clrWizardPageNextDisabledChanged",] },],
    };
    return WizardPage;
}(TabContent));
